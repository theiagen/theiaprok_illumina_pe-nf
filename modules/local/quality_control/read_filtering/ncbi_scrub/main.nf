process NCBI_SCRUB_PE {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/theiagen/sra-human-scrubber:2.2.1"

    input:
    tuple val(meta), path(read1), path(read2)

    output:
    tuple val(meta), path("*_dehosted.fastq.gz"), emit: dehosted_reads
    tuple val(meta), path("SPOTS_REMOVED.txt"), emit: spots_removed_file
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    """
    # date and version control
    date | tee DATE
    
    # detect compression and define cat command
    if [[ "${read1}" == *.gz ]]; then
        echo "DEBUG: Gzipped input reads detected"
        cat_command="zcat"
    else
        cat_command="cat"
    fi
    
    # Count the number of reads in each file
    read1_count=\$(\$cat_command ${read1} | wc -l | awk '{print \$1/4}')
    read2_count=\$(\$cat_command ${read2} | wc -l | awk '{print \$1/4}')
    echo "DEBUG: Number of reads in read1: \$read1_count"
    echo "DEBUG: Number of reads in read2: \$read2_count"
    
    if [[ \$read1_count -ne \$read2_count ]]; then
        echo "ERROR: The number of reads in the two input files do not match."
        echo "ERROR: The number of reads in read1 is \$read1_count and the number of reads in read2 is \$read2_count."
        echo "ERROR: The unpaired reads will be ignored from the interleaved file..."
    fi
    
    # interleave reads
    # paste command takes 4 lines at a time and merges them into a single line with tabs
    # tr substitutes the tab separators from paste into new lines, effectively interleaving the reads and keeping the FASTQ format
    echo "DEBUG: Interleaving reads with paste..."
    paste <(\$cat_command ${read1} | paste - - - -) <(\$cat_command ${read2} | paste - - - -) | tr '\\t' '\\n' > interleaved.fastq
    
    # dehost reads
    # -x Remove spots instead of default 'N' replacement.
    # -s Input is (collated) interleaved paired-end(read) file AND you wish both reads masked or removed.
    echo "DEBUG: Running HRRT..."
    echo "DEBUG: /opt/scrubber/scripts/scrub.sh -p ${task.cpus} -x -s -i interleaved.fastq"
    /opt/scrubber/scripts/scrub.sh -p ${task.cpus} -x -s -i interleaved.fastq > STDOUT 2> STDERR
    
    # capture the number of spots removed by fetching the first item on the last line of stderr
    spots_removed=\$(cat STDERR | tail -n1 | awk -F" " '{print \$1}')
    echo "\$spots_removed" > SPOTS_REMOVED.txt
    echo "DEBUG: Human spots removed: \$spots_removed"
    
    # split interleaved reads and compress files
    # paste takes 8 lines at a time and merges them into a single line with tabs
    # grouping each pair of reads into one line
    echo "DEBUG: Splitting interleaved dehosted reads..."
    paste - - - - - - - - < interleaved.fastq.clean \\
        | tee >(cut -f 1-4 | tr '\\t' '\\n' | gzip > ${prefix}_R1_dehosted.fastq.gz) \\
        | cut -f 5-8 | tr '\\t' '\\n' | gzip > ${prefix}_R2_dehosted.fastq.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ncbi-scrubber: \$(echo "2.2.1")
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_R1_dehosted.fastq.gz
    touch ${prefix}_R2_dehosted.fastq.gz
    echo "0" > SPOTS_REMOVED.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ncbi-scrubber: 2.2.1
    END_VERSIONS
    """
}