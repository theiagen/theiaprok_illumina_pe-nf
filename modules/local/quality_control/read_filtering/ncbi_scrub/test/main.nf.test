nextflow_process {
    name "Test Process NCBI_SCRUB_PE"
    script "../main.nf"
    process "NCBI_SCRUB_PE"
    
    test("ncbi_scrub test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.read1_dehosted,
                    process.out.read2_dehosted,
                    process.out.spots_removed_file,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("ncbi_scrub - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}