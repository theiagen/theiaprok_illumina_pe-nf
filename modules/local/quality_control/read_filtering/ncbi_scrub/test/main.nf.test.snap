{
    "ncbi_scrub - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "read1_dehosted": [
                    
                ],
                "read2_dehosted": [
                    
                ],
                "spots_removed_file": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-17T20:01:54.730794"
    },
    "ncbi_scrub test": {
        "content": [
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-17T20:01:42.957432"
    }
}